### PR TITLE
*: support history read with old schema

### PR DESCRIPTION
**Pull request by coocood**
When user set 'tidb_snapshot' variable, load and use the schema at the snapshot time.

### CODE REVIEW

File: domain/domain.go [hunk 1]
```diff
@@ -44,26 +44,20 @@ type Domain struct {
...
 	if err != nil {
 		return errors.Trace(err)
 	}
-
-	info := do.infoHandle.Get()
-	if info != nil && schemaMetaVersion <= info.SchemaMetaVersion() {
-		// info may be changed by other txn, so here its version may be bigger than schema version,
-		// so we don't need to reload.
-		log.Debugf("[ddl] schema version is still %d, no need reload", schemaMetaVersion)
+	m := meta.NewSnapshotMeta(snapshot)
```
Comment (shenli) [inline domain/domain.go]: Why not create a meta outside this function and pass is as an argument?

File: executor/executor_simple.go [hunk 2]
```diff
@@ -203,11 +203,32 @@ func (e *SimpleExec) executeSet(s *ast.SetStmt) error {
...
 	}
 	return nil
 }
 
+func (e *SimpleExec) loadSnapshotInfoSchemaIfNeeded(sessionVars *variable.SessionVars) error {
+	if sessionVars.SnapshotTS == 0 {
+		sessionVars.SnapshotInfoschema = nil
+		return nil
+	}
+	log.Infof("loadSnapshotInfoSchema")
```
Comment (shenli) [inline executor/executor_simple.go]: Add  SnapshotTS into log.

File: executor/executor_simple.go [hunk 3]
```diff
@@ -203,11 +203,32 @@ func (e *SimpleExec) executeSet(s *ast.SetStmt) error {
 			if err != nil {
 				return errors.Trace(err)
 			}
+			if name == "tidb_snapshot" {
```
Comment (shenli) [inline executor/executor_simple.go]: Make "tidb_snapshot" a const.

**Comment by c4pt0r** [comment]
this PR is out-of-date, PTAL @coocood

**Comment by coocood** [comment]
@c4pt0r I will update branch when it's ready to merge.

File: domain/domain.go (hunk 1 above)
Comment (coocood) [inline domain/domain.go]: I tried to extract create meta outside of this function, but it adds many lines of code.
create meta inside, so the caller doesn't need to care about meta.

Comment (shenli) [inline domain/domain.go]: You should add comments for the function.

File: domain/domain.go [hunk 4]
```diff
@@ -103,6 +96,16 @@ func (do *Domain) InfoSchema() infoschema.InfoSchema {
 	return do.infoHandle.Get()
 }
 
+// GetSnapshotInfoSchema gets a snapshot information schema.
+func (do *Domain) GetSnapshotInfoSchema(snapshotTS uint64) (infoschema.InfoSchema, error) {
```
Comment (zimulala) [inline domain/domain.go]: Why not use `MustReload` directly？ If use this function it needs to set `SchemaValidity` value.

**Comment by zimulala** [comment]
LGTM

**Comment by shenli** [comment]
LGTM

**Pull request merged**